File: SortingAlgorithm/_8_RadixSort/RadixSort.java
package Algorithm.SortingAlgorithm._8_RadixSort;
import java.util.Arrays;

class RadixSort {
    public static int[] radixSort(int[] arr) {
        int n = arr.length;
        int max = Arrays.stream(arr).max().getAsInt();

        for (int exp = 1; max / exp > 0; exp *= 10) {
            arr = countSort(arr, n, exp);
        }

        return arr;
    }

    public static int[] countSort(int[] arr, int n, int exp) {
        int[] output = new int[n];
        int[] count = new int[10];

        Arrays.fill(count, 0);

        for (int i = 0; i < n; i++) {
            count[(arr[i] / exp) % 10]++;
        }

        for (int i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }

        for (int i = n - 1; i >= 0; i--) {
            output[count[(arr[i] / exp) % 10] - 1] = arr[i];
            count[(arr[i] / exp) % 10]--;
        }

        System.arraycopy(output, 0, arr, 0, n);

        return arr;
    }

    public static void main(String[] args) {
        int[] arr = {170, 45, 75, 90, 802, 24, 2, 66};
        int n = arr.length;

        int[] sortedArray = radixSort(arr);

        System.out.println("Sorted Array: " + Arrays.toString(sortedArray));
    }
}

File: SortingAlgorithm/_8_RadixSort/RadixSort.py
def countingSort(arr, exp):
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    # Count the occurrences of each digit at the current position
    for i in range(n):
        index = (arr[i] // exp) % 10
        count[index] += 1

    # Calculate the actual position of each digit in the output array
    for i in range(1, 10):
        count[i] += count[i - 1]

    # Build the output array
    i = n - 1
    while i >= 0:
        index = (arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1
        i -= 1

    # Copy the sorted output back to the original array
    for i in range(n):
        arr[i] = output[i]


def radixSort(arr):
    max_num = max(arr)
    exp = 1

    while max_num // exp > 0:
        countingSort(arr, exp)
        exp *= 10


# Driver code
arr = [170, 45, 75, 90, 802, 24, 2, 66]
radixSort(arr)

print("Sorted Array:", arr)
